testsuites/store/src/main/java/org/eclipse/rdf4j/sail/SailConcurrencyTest.java

```diff
@@ -127,7 +127,7 @@ public void run() {
 						conn.addStatement(subject, RDFS.COMMENT, vf.createLiteral("ci" + i), context);
 					}
 					if (rollback) {
-						otherTxnCommitted.await();
+						otherTxnCommitted.await(); // deadlocks here. Waiting for another transaction while holding the exclusive write lock!
 						conn.rollback();
```

Comment:
The test itself is correct, but it assumes that the sail stack supports concurrent writes. Like @ansell said: I think you should disable this test for your reasoner, and document the fact that it does not fully support concurrent transactions.